In [36]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
import matplotlib as plt
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import warnings
#To ignore all warnings
warnings.filterwarnings('ignore')

In [37]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [38]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [39]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [40]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
# Scaling the numeric columns
market_data_scaled_values = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d",	"price_change_percentage_30d", 
                                                                    "price_change_percentage_60d",	"price_change_percentage_200d",	"price_change_percentage_1y"]])

market_data_scaled_values[0:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [41]:

# Create a DataFrame with the scaled data
market_data_scaled = pd.DataFrame(
    market_data_scaled_values, 
    columns=[
    "price_change_percentage_24h", 
    "price_change_percentage_7d", 
    "price_change_percentage_14d",	
    "price_change_percentage_30d", 
    "price_change_percentage_60d",	
    "price_change_percentage_200d",	
    "price_change_percentage_1y"
    ], 
    index=df_market_data.index
)

# Review the first few rows of the updated DataFrame
market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [42]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))

In [43]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data)
    inertia.append(k_model.inertia_)


In [71]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame using the elbow_data Dictionary
elbow_original = pd.DataFrame(elbow_data)

# Review the DataFrame
elbow_original.head()

#elbow_data


k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [72]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_original.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  The curve comes to a point at the number 4 tick-mark and then the curve flattens, because of this the best value for k is 3.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [73]:
# Initialize the K-Means model using the best value for k
# Define the model with optimal number of clusters
model = KMeans(n_clusters=4)

In [74]:
# Fit the K-Means model using the scaled data
model.fit(market_data_scaled)

KMeans(n_clusters=4)

In [75]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_clusters = model.predict(market_data_scaled)

# Print the resulting array of cluster values.
k_clusters

array([3, 3, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 3, 0, 3, 0, 0, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 2, 3, 0, 0, 1, 0, 0, 0, 0])

In [76]:
# Create a copy of the DataFrame
crypto_predictions_df = market_data_scaled.copy()

In [77]:

# Add a new column to the DataFrame with the predicted clusters
crypto_predictions_df['clusters'] = k_clusters

# Display sample data
crypto_predictions_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters  
coin_id                                             
bitcoin                        -0.251637         3  
ethereum                       -0.199352         3  
tether                         -0.282061         0  
ripple                         -0.295546         0  
bitcoin-cash                   -0.270317         3

In [78]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
crypto_predictions_df.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters",
    hover_cols=["crypto_name"],  # Add crypto name to hover information
    title="Cryptocurrency Price Changes: 24h vs 7d"
)

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimize Clusters with Principal Component Analysis.

In [79]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [80]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

market_pca = pca.fit_transform(market_data_scaled)

# View the first five rows of the DataFrame. 
market_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [81]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [82]:
x=0.3719856+0.34700813+0.17603793

print(x)

0.89503166


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:**  89.50%

In [83]:
# Create a new DataFrame with the PCA data.
# Copy the crypto names from the original data
# Set the coinid column as index

market_pca_df = pd.DataFrame(
    market_pca,
    columns=["PCA1", "PCA2", "PCA3"],
    index=df_market_data.index
)

# Review the PCA DataFrame
market_pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [84]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [85]:
# Create an empty list to store the inertia values
inertia=[]

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(market_pca_df)
    inertia.append(k_model.inertia_)


In [86]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame using the elbow_data Dictionary
elbow_pca = pd.DataFrame(elbow_data)

# Review the DataFrame
elbow_pca.head()


k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [87]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**  4 is the best value for k


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:**  No.  The Original Data also showed that 4 was the best k value.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [88]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4)

In [89]:
# Fit the K-Means model using the PCA data
model.fit(market_pca_df)

KMeans(n_clusters=4)

In [90]:

# Predict the clusters to group the cryptocurrencies using the scaled data
k_clusters = model.predict(market_pca_df)

# Print the resulting array of cluster values.
k_clusters

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 2, 0, 1, 1, 3, 1, 1, 1, 1])

In [92]:

# Create a copy of the DataFrame
crypto_pca_predictions_df = market_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
crypto_pca_predictions_df['clusters'] = k_clusters

# Display sample data
crypto_pca_predictions_df.head()

PCA1      PCA2      PCA3  clusters
coin_id                                             
bitcoin      -0.600667  0.842760  0.461595         0
ethereum     -0.458261  0.458466  0.952877         0
tether       -0.433070 -0.168126 -0.641752         1
ripple       -0.471835 -0.222660 -0.479053         1
bitcoin-cash -1.157800  2.041209  1.859715         0

In [93]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

crypto_pca_predictions_df.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="clusters",
    hover_cols=["crypto_name"],  # Add crypto name to hover information
)

:NdOverlay   [clusters]
   :Scatter   [PCA1]   (PCA2)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [94]:
# Check individual plots
elbow_pca_plot = elbow_pca.hvplot.line(x='k', y='inertia', label='PCA Data', color='orange')
elbow_original_plot = elbow_original.hvplot.line(x='k', y='inertia', label='Original Data', color='blue')

elbow_pca_plot.opts(title="PCA Data Elbow Curve")
#elbow_original_plot.opts(title="Original Data Elbow Curve")

:Curve   [k]   (inertia)

In [95]:
elbow_original_plot.opts(title="Original Data Elbow Curve")

:Curve   [k]   (inertia)

In [96]:
elbow_original_plot + elbow_pca_plot


:Layout
   .Curve.Original_Data :Curve   [k]   (inertia)
   .Curve.PCA_Data      :Curve   [k]   (inertia)

In [97]:

# Plot clusters for original data
clusters_original_plot = crypto_predictions_df.hvplot.scatter(x='price_change_percentage_24h', y='price_change_percentage_7d', by='clusters', colormap='Category10', title='Original Data Clusters')

# Plot clusters for PCA data
clusters_pca_plot = crypto_pca_predictions_df.hvplot.scatter(x='PCA1', y='PCA2', by='clusters', colormap='Category10', title='PCA Data Clusters')


# Composite plot
composite_clusters = clusters_original_plot + clusters_pca_plot
composite_clusters

:Layout
   .NdOverlay.I  :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .NdOverlay.II :NdOverlay   [clusters]
      :Scatter   [PCA1]   (PCA2)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Impact of Using Fewer Features in Clustering

When using fewer features (via PCA) to cluster the data: 

**Simplicity:** The clusters are more distinct and easier to interpret due to reduced dimensionality.

**Potential Loss of Information:** Some nuances in the data could be lost, leading to slightly different cluster boundaries or even misclassification.

**Computational Efficiency:** PCA reduces the complexity, making clustering faster and less resource-intensive.

**The visual comparison of clusters reflect these trade-offs, showing that the PCA-transformed data clusters overlap more but also have a few spread out clusters.**